### PR TITLE
[JENKINS-27058] Clarify that remoteFS can be relative if you know what y...

### PR DESCRIPTION
**Pull request by stephenc**
...ou are doing
- Get the absolute path on connect and use that while the connection is on-line (which is ok as when the connection is not online a null was returned from getRootPath() anyway)

### CODE REVIEW

**Comment by stephenc** [comment]
To replace https://github.com/jenkinsci/jenkins/pull/1578 @reviewbybees

File: core/src/main/java/hudson/model/Slave.java [hunk 1]
```diff
@@ -121,14 +126,14 @@
      * Whitespace-separated labels.
      */
     private String label="";
-    
+
     private /*almost final*/ DescribableList<NodeProperty<?>,NodePropertyDescriptor> nodeProperties = new DescribableList<NodeProperty<?>,NodePropertyDescriptor>(Jenkins.getInstance());
```
Comment (jtnord) [inline core/src/main/java/hudson/model/Slave.java]: people in greenhouses shouldn't throw stones :-)

Comment (stephenc) [inline core/src/main/java/hudson/model/Slave.java]: I comment on other's whitespace changes iff they have been especially nit-picky in their own review comments

**Comment by jtnord** [comment]
:+1: ignoring whitespace unrelated changes.

File: core/src/main/java/hudson/model/Slave.java [hunk 2]
```diff
@@ -425,7 +437,7 @@ public FormValidation doCheckRemoteFS(@QueryParameter String value) throws IOExc
             if (!value.contains("\\") && !value.startsWith("/")) {
                 // Unix-looking path that doesn't start with '/'
                 // TODO: detect Windows-looking relative path
-                return FormValidation.error(Messages.Slave_the_remote_root_must_be_an_absolute_path());
+                return FormValidation.error(Messages.Slave_Remote_Relative_Path_Warning());
```
Comment (jglick) [inline core/src/main/java/hudson/model/Slave.java]: Not `.warning`?

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 3]
```diff
@@ -410,6 +412,11 @@ public long getResourceLoadingTime() throws IOException, InterruptedException {
         return channel.call(new LoadingTime(true));
     }
 
+    @CheckForNull
+    public String getAbsoluteRemoteFs() {
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: Either `@Restricted` or `@since` or not `public`.

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 4]
```diff
@@ -481,11 +488,16 @@ public void onClosed(Channel c, IOException cause) {
...
             throw new IOException("Node "+nodeName+" has been deleted during the channel setup");
         }
-        
-        String remoteFs = node.getRemoteFS();
-        if(_isUnix && !remoteFs.contains("/") && remoteFs.contains("\\"))
-            log.println("WARNING: "+remoteFs+" looks suspiciously like Windows path. Maybe you meant "+remoteFs.replace('\\','/')+"?");
-        FilePath root = new FilePath(channel,remoteFs);
+
+        String absoluteRemoteFs = node.getRemoteFS();
+        if (!absoluteRemoteFs.startsWith("\\") && !absoluteRemoteFs.startsWith("/")) {
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: What about `C:\absolute`?

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 5]
```diff
@@ -634,9 +647,13 @@ public RetentionStrategy getRetentionStrategy() {
      */
     private void closeChannel() {
         // TODO: race condition between this and the setChannel method.
-        Channel c = channel;
-        channel = null;
-        isUnix = null;
+        Channel c;
+        synchronized (channelLock) {
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: How is this related to the rest of the PR?

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 6]
```diff
@@ -708,6 +725,18 @@ public Boolean call() throws IOException {
         }
     }
 
+    private static final class AbsolutePath extends MasterToSlaveCallable<String,IOException> {
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: `serialVersionUID`

File: core/src/main/resources/hudson/model/Messages.properties [hunk 7]
```diff
@@ -244,7 +244,9 @@ Slave.Launching={0} Launching slave agent
 Slave.Network.Mounted.File.System.Warning=Are you sure you want to use network mounted file system for FS root? Note that this directory does not need to be visible to the master.
 Slave.Remote.Director.Mandatory=Remote directory is mandatory
 Slave.Terminated={0} slave agent was terminated
-Slave.the_remote_root_must_be_an_absolute_path=The remote root must be an absolute path.
+Slave.Remote.Relative.Path.Warning=Are you sure you want to use a relative path for the FS root? Note that relative \
+  paths require that you can assure that the selected launcher provides a consistent current working directory. Using \
+  an absolute path is strongly encouraged and highly recommended.
```
Comment (jglick) [inline core/src/main/resources/hudson/model/Messages.properties]: `strongly encouraged and highly recommended` is redundant.

**Comment by jglick** [comment]
Hard to review given the excessive whitespace changes, but :+1: overall insofar as I understand the implications.

File: core/src/main/resources/hudson/model/Messages.properties [hunk 8]
```diff
@@ -244,7 +244,9 @@ Slave.Launching={0} Launching slave agent
 Slave.Network.Mounted.File.System.Warning=Are you sure you want to use network mounted file system for FS root? Note that this directory does not need to be visible to the master.
 Slave.Remote.Director.Mandatory=Remote directory is mandatory
 Slave.Terminated={0} slave agent was terminated
-Slave.the_remote_root_must_be_an_absolute_path=The remote root must be an absolute path.
+Slave.Remote.Relative.Path.Warning=Are you sure you want to use a relative path for the FS root? Note that relative \
+  paths require that you can assure that the selected launcher provides a consistent current working directory. Using \
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Messages.properties]: What does this mean?

Comment (jglick) [inline core/src/main/resources/hudson/model/Messages.properties]: I think Stephen is trying to say something about cloudy slaves that share a common filesystem despite apparently different launch details, but I am not really sure either. Deserves a clearer explanation.

Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Messages.properties]: This should be part of the documentation help text.

**Comment by stephenc** [comment]
@daniel-beck @jglick I think I have addressed your issues

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 9]
```diff
@@ -537,6 +559,22 @@ public void onClosed(Channel c, IOException cause) {
         Jenkins.getInstance().getQueue().scheduleMaintenance();
     }
 
+    /*package*/ static boolean isRelativePath(String path) {
+        if (path.startsWith("/"))
+            return false;
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: Debatable whether you want to consider `\Windows\Temp\slave` absolute or relative. Technically it is drive-relative.

File: core/src/test/java/hudson/slaves/SlaveComputerTest.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        assertThat("\\", aRelativePath());
+        assertThat("\\foo\\bar", aRelativePath());
+        assertThat("\\foo\\..\\bar", aRelativePath());
+        assertThat("", aRelativePath());
+        assertThat(".", aRelativePath());
+        assertThat(".\\foo", aRelativePath());
+        assertThat(".\\foo\\bar", aRelativePath());
+        assertThat(".\\foo\\bar\\", aRelativePath());
+        assertThat("\\\\foo", aRelativePath());
+        assertThat("\\\\foo\\", not(aRelativePath()));
```
Comment (jglick) [inline core/src/test/java/hudson/slaves/SlaveComputerTest.java]: FWIW the share name is a mandatory part of a UNC path, so `\\server\share\localpath`. I doubt the distinction matters here.

File: core/src/test/java/hudson/slaves/SlaveComputerTest.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        assertThat("c:.", aRelativePath());
+        assertThat("c:\\", not(aRelativePath()));
+        assertThat("c:/", not(aRelativePath()));
+    }
+
+    private static RelativePathMatcher aRelativePath() {
+        return new RelativePathMatcher();
+    }
+
+    private static class RelativePathMatcher extends BaseMatcher<String> {
```
Comment (jglick) [inline core/src/test/java/hudson/slaves/SlaveComputerTest.java]: Does Hamcrest actually make your life easier?

**Comment by jglick** [comment]
:+1:

**Comment by jglick** [comment]
Recommend review from @recampbell since he encountered the issue to begin with.

File: core/src/test/java/hudson/slaves/SlaveComputerTest.java (hunk 11 above)
Comment (stephenc) [inline core/src/test/java/hudson/slaves/SlaveComputerTest.java]: I prefer reading it this way, so yes

File: core/src/test/java/hudson/slaves/SlaveComputerTest.java (hunk 10 above)
Comment (stephenc) [inline core/src/test/java/hudson/slaves/SlaveComputerTest.java]: I know but I wanted a faster test and this is better than anything else we had... it can be improved if somebody feels the need ;-)

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 9 above)
Comment (stephenc) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: `\windows\temp\slave` is relative, `/windows/temp/slave` is fine to consider as absolute as you are indicating that you think it is unix and the `isUnix` test will warn you about the path being for a different OS

**Comment by stephenc** [comment]
I'll wait for @recampbell so...

**Comment by recampbell** [comment]
:+1:

**Comment by stephenc** [comment]
Jenkins found a failing test with the changes for @jglick, so some refactoring to reduce code duplication

**Comment by stephenc** [comment]
Test failures in most recent build are irrelevant

**Comment by KostyaSha** [comment]
Why all whitespace changes were done in one commit with real changes?

**Comment by stephenc** [comment]
IntelliJ screwed me over and refused to unscrew

**Comment by KostyaSha** [comment]
On left side of the code (near line numbers) there is a grey colour that shows that something was changed. If you press on it it allows to undo it. Also there was somewhere diff button that allows compare current vs previous changes and merge them.

**Comment by jglick** [comment]
At worst,

1. `git checkout mybranch`
1. `git merge master`
1. `git diff master > /tmp/x.diff`
1. edit `/tmp/x.diff` and delete all useful changes
1. `git apply -R /tmp/x.diff`
1. `git commit -A -m 'Writing ten times on the blackboard, I will not commit gratuitous WS changes again.'`

:-)

**Comment by KostyaSha** [comment]
@jglick just `git add -i ` then `patch` and skip WS hanks.

**Comment by stephenc** [comment]
@KostyaSha so the issue I had is that I set intellij to leave whitespace alone on unmodified lines, but then every time I did the grey trick it would revert it on save... Normally I don't have this issue but I think it was seeing me "undoing" the whitespace changes as changing those lines and thus they were up for whitespace fixing... lather rinse repeat

**Comment by KostyaSha** [comment]
I think we should define settings and document on wiki, i also spending much time doing PRs without whitespaces :(

**Comment by KostyaSha** [comment]
screenshot 2015-03-23 21 46 12
Seems this IDEA 14 feature doesn't work right. It acts like "All"

**Comment by KostyaSha** [comment]
Reproduces, open SafeTimerTask.class, add new line, alt+tab window, back... May this be caused by some git configuration?

**Comment by stephenc** [comment]
That's the exact issue I had. It used to work in IDEA 13... ffs

**Comment by KostyaSha** [comment]
Filled bug request to their support because i have no other ideas.

**Comment by KostyaSha** [comment]
@stephenc support replied, IDEA supports editor config https://youtrack.jetbrains.com/issue/IDEA-130042#comment=27-804824 that was put by this https://github.com/jenkinsci/jenkins/pull/1310  https://github.com/jenkinsci/jenkins/blob/master/.editorconfig#L16 
As i see this format sucks on options and my proposal to remove it.

**Comment by jtnord** [comment]
it's also has no support for eclipse :-)

Note: even eclipse for the same eclipse format config will generate
different formats between versions as bugs get fixed/introduced and new
features are added.

**Comment by KostyaSha** [comment]
https://github.com/jenkinsci/jenkins/pull/1619
